extension/api.js

```diff
@@ -78,14 +105,174 @@ const lostEngines = new Map ([
   ["{39790485-930b-40a5-8268-69222363ff80}", Oct2020],
   ["{446c7519-9e32-4f9a-b562-447f4421ec9a}", Oct2020],
   ["{b30e775a-7a10-480d-ace4-761b9ca07aee}", Dec2020],  // Test add-on
+  // ["{fd299ce1-1602-4490-b659-f45504f9324c}", Date.now()], // Manual test addon for dev.
 ]);
 
+// If the extension is not running or enabled in the search service
+// we will try again later.
+let tryagain = false;
+
+function finish(result = false) {
+  Services.prefs.setBoolPref(RUN_ONCE_PREF, result);
+}
+
+async function tryDefaultReset() {
+  let defaultEngine = await Services.search.getDefault();
+  if (lostEngines.has(defaultEngine?._extensionID)) {
+    console.log(`reset-default-search: The default engine was already selected by the user. ${defaultEngine._extensionID}`);
+    finish(true);
+    return;
+  }
+
+  // Make sure the setting is not user controlled.
+  let control = await ExtensionSettingsStore.getLevelOfControl(
+    ExtensionSettingsStore.SETTING_USER_SET,
+    DEFAULT_SEARCH_STORE_TYPE,
+    DEFAULT_SEARCH_SETTING_NAME
+  );
+  if (control == "controlled_by_this_extension") {
+    console.log(`reset-default-search: search control has been user selected, not resetting, finished.`)
+    finish();
+    return;
+  }
+
+  // Get the latest installed addon that was installed prior to it's failure date.  Any non-app
+  // disabled state (ie. it is not disabled due to blocklisting) removes the addon from elibility.
+  let addons = (await AddonManager.getAddonsByIDs(
+      Array.from(lostEngines.keys())
+    )).filter(
+      a => a &&
+      !a.userDisabled &&
+      !a.softDisabled &&
+      !a.embedderDisabled &&
+      lostEngines.has(a.id) &&
+      a.installDate < lostEngines.get(a.id)
+    );
+  if (!addons.length) {
+    console.log("reset-default-search: No addons in our list are installed.");
+    finish();
+    return;
+  }
+
+  // Filter out any that are blocklisted.
+  let enabledAddons = addons.filter(a => !a.appDisabled);
+  console.log(`enabledAddons ${enabledAddons.length}`);
```

Comment:
Is this log useful? If so, prepend `reset-default-search: ` in front of it for consistency. If not, remove the line.

Reply:
Right now, the more logging to indicate issues the better.